tachyon/math/matrix/matrix_utils.h

```diff
@@ -76,6 +79,50 @@ std::vector<PackedPrimeField> PackRowVertically(
   });
 }
 
+// Expands a |RowMajorMatrix|'s rows from |rows| to |rows|^{|added_bits|},
+// moving values from row |i| to row |i|^{|added_bits|}. All new entries are set
+// to |F::Zero()|.
+template <typename Derived>
+void ExpandInPlaceWithZeroPad(Eigen::MatrixBase<Derived>& mat,
+                              size_t added_bits) {
+  if (added_bits == 0) {
+    return;
+  }
+
+  Eigen::Index original_rows = mat.rows();
+  Eigen::Index new_rows = mat.rows() << added_bits;
+  Eigen::Index cols = mat.cols();
+
+  Derived padded = Derived::Constant(new_rows, cols, Derived::Scalar::Zero());
+
+  OPENMP_PARALLEL_FOR(Eigen::Index row = 0; row < original_rows; ++row) {
+    Eigen::Index padded_row_index = row << added_bits;
+    // TODO(ashjeong): Check if moved properly
+    padded.row(padded_row_index) = std::move(mat.row(row));
+  }
+  mat = std::move(padded);
+}
+
+// Swaps rows of a |RowMajorMatrix| such that each row is changed to the row
```

Comment:
Shouldn't we add a check if the matrix is row major like PackRowHorizontally?

Reply:
This function became generalized and no longer only pertains to `RowMajorMatrices`. I'll fix the comments!